### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by freva**
This should fix the issues outlined in #116.

I had trouble getting stable and accurate metrics for memory usage, but I did include metrics for runtime execution. All tests results are average milliseconds of 5 runs with 1 warmup round that is not counted.

100k random tweets:

Method | e867585 | 6bc16b7 | Δ e867585 | 25a6d91 | Δ e867585
-- | -- | -- | -- | -- | --
parseToAliases | 228.63 | 232.88 | 1.86% | 128.76 | -43.68%
parseToHtmlDecimal | 234.51 | 224.61 | -4.22% | 126.83 | -45.91%
parseToHtmlHexadecimal | 239.59 | 225.80 | -5.76% | 125.51 | -47.62%
removeAllEmojis | 234.01 | 214.23 | -8.45% | 128.81 | -44.96%
parseToUnicode | 6611.55 | 99.26 | -98.50% | 86.75 | -98.69%

Same 100k tweets from previous test, but with ~350k HTML encoded emojis inserted into ~83k/100k of the tweets. 

Method | e867585 | 6bc16b7 | Δ e867585 | 25a6d91 | Δ e867585
-- | -- | -- | -- | -- | --
parseToAliases | 349.58 | 343.24 | -1.81% | 191.25 | -45.29%
parseToHtmlDecimal | 345.90 | 384.49 | 11.16% | 193.41 | -44.09%
parseToHtmlHexadecimal | 336.58 | 357.78 | 6.30% | 213.58 | -36.54%
removeAllEmojis | 354.79 | 329.84 | -7.03% | 210.94 | -40.55%
parseToUnicode | 10438.70 | 4721.01 | -54.77% | 209.65 | -97.99%

To sum up, it speeds up `parseToUnicode()` by about 50 times and the four from-unicode methods by a factor of 2. The improvement for the from-unicode methods is mostly due to avoiding copying the char array when searching for emoji end.

### CODE REVIEW

**Comment by vdurmont** [comment]
😮 😍 🚀 👍
